[pyper][emb][quantization] Support emb trained in FP16

ShijunK opened the pull request:
Summary: Add support of embedding with input data type as float16, utilize new kernel functions added in fbgemm https://github.com/pytorch/FBGEMM/pull/616

Test Plan: `buck test caffe2/test/:quantization -- test_embedding_bag`

Differential Revision: D29392320

supriyar commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:
```diff
@@ -299,16 +312,34 @@ Tensor _qembeddingbag_nbit_prepack_helper(
+
   } else {
 #endif // USE_FBGEMM
+    TORCH_CHECK(weight.scalar_type() == at::ScalarType::Float, "'embedding_bag_byte_prepack' only support float32, when not use fbgemm.");
```
Can we add a fallback path for fp16 as well that works without fbgemm? It can just cast the input and do the regular computaion

ShijunK replied:
pushed a new version with fallback logic (cast fp16 to fp32) when fbgemm is not enabled.

supriyar commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:
```diff
@@ -299,28 +323,45 @@ Tensor _qembeddingbag_nbit_prepack_helper(
 
     for (int row = 0; row < embedding_rows; ++row) {
-      const float* input_row = weight_data + row * embedding_cols;
+      const auto row_tensor = weight_contig.scalar_type() == at::ScalarType::Half
+        ? weight_contig[row].to(at::ScalarType::Float)
+        : weight_contig[row];
```
Can we move the casting outside the for-loop? similar to `qembeddingbag_byte_prepack` above?

supriyar commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:
```diff
@@ -262,12 +283,15 @@ Tensor _qembeddingbag_nbit_prepack_helper(
     const double ratio) {
+  TORCH_CHECK(
+    weight.scalar_type() == at::ScalarType::Float || weight.scalar_type() == at::ScalarType::Half,
+    "'embedding_bag_byte_prepack' only support float32 or float16.");
```
Can we  update the message here for 4/2 bit?

supriyar reviewed (approved):
Looks good, please address last couple comments before landing

ShijunK committed:
[pyper][emb][quantization] Support emb trained in FP16 (#60736)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/60736

Add support of embedding with input data type as float16, utilize new kernel functions added in fbgemm https://github.com/pytorch/FBGEMM/pull/616

Test Plan: `buck test caffe2/test/:quantization -- test_embedding_bag`

Reviewed By: supriyar

Differential Revision: D29392320

fbshipit-source-id: bbd110f78e475a00f2b33ceed6958ddafbf7bfaa
